[#718] CSR and Private Key Generation + Storage

sam-bertin opened the pull request:
Taiga ticket => https://taiga.maykinmedia.nl/project/maykin-intranet/issue/718

sergei-maertens commented on CHANGELOG.rst:
```diff
@@ -2,6 +2,14 @@
+2.1.0
+=====
+
+*July 12, 2024*
```
please do not add release dates to feature PRs, or even the version number. you can put something like `dev (unreleased)` instead for the version, but usually I generate the changelog from the commits when we decide to create a new release

sergei-maertens commented on README.rst:
```diff
@@ -1,7 +1,7 @@
 =========================
 
-:Version: 2.0.0
+:Version: 2.1.0
```
please don't make these changes in non-release branches

sergei-maertens commented on docs/conf.py:
```diff
@@ -30,7 +30,7 @@
 
 # The full version, including alpha/beta/rc tags
-release = "2.0.0"
+release = "2.1.0"
```
same as above

sergei-maertens commented on docs/quickstart.rst:
```diff
@@ -34,6 +34,12 @@ configuration.
 Whenever an instance is deleted (through the admin or code), the associated files are
 purged as well.
 
+You can now also generate a Certificate Signing Request (CSR) and a private key in the admin interface.
+
+Generated files are stored at your configured ``PRIVATE_ROOT`` directory.
+
+You can also download the CSR from the admin interface.
```
drop the "you can now" - documentation itself is also versioned should it should just describe the $current situation :)

sergei-maertens commented on pyproject.toml:
```diff
@@ -4,7 +4,7 @@ build-backend = "setuptools.build_meta"
 [project]
 name = "django-simple-certmanager"
-version = "2.0.0"
+version = "2.1.0"
```
let's keep this out of the feature PR

sergei-maertens commented on pyproject.toml:
```diff
@@ -31,6 +31,8 @@ dependencies = [
     "django-privates>=1.5",
     "cryptography>=35.0.0",
+    "types-pyOpenSSL>=24.1.0",
+    "pyOpenSSL>=24.1.0",
```
we just dropped this dependency in #27 - you should be able to do all this with the `cryptography` library.

sergei-maertens commented on simple_certmanager/models.py:
```diff
@@ -98,3 +116,35 @@ def is_valid_key_pair(self) -> None | bool:
+    organization_name = models.CharField(
+        _("organization name"),
+        max_length=100,
+        help_text=_("The name of the organization, e.g. 'Gementee Amsterdam'"),
```
some organizations will not appreciate that "Amsterdam" is here, let's use a fictive organization name:

```suggestion
        help_text=_("The name of the organization, e.g. 'OpenGem'"),
```

sergei-maertens commented on simple_certmanager/models.py:
```diff
@@ -43,6 +47,20 @@ class Certificate(DeleteFileFieldFilesMixin, models.Model):
         upload_to="ssl_certs_keys/%Y/%m/%d",
         validators=[PrivateKeyValidator()],
     )
+    csr = PrivateMediaFileField(
+        _("certificate signing request"),
+        help_text=_("The content of the certificate signing request"),
+        upload_to="ssl_certs_keys/%Y/%m/%d",
+        blank=True,
+        validators=[CertificateSigningRequestValidator()],
+    )
```
I don't think this field should exist on the `Certificate` model - it's only relevant in the request phase but not needed for runtime at all.

Instead, I'd rather add a `models.TextField` on the `CertificateSigningRequest` model where the content of the CSR is stored, that can then easily be copy-pasted for the actual request (and you can add a convenience "download" link in the admin to generate a file with these contents).

Once the certificate itself is added, there's no need to keep the CSR around anymore, so let's just not clutter the file system with them.

CharString replied:
If compute allows it, CSR could even just be generated on request and not stored at all.

sergei-maertens commented on simple_certmanager/models.py:
```diff
@@ -43,6 +47,20 @@ class Certificate(DeleteFileFieldFilesMixin, models.Model):
+        blank=True,
+        validators=[CertificateSigningRequestValidator()],
+    )
+    info = models.OneToOneField(
+        "CertificateSigningRequestInfo",
+        on_delete=models.CASCADE,
+        verbose_name=_("certificate information"),
+        help_text=_("Information about the certificate"),
+        null=True,
+    )
```
I'd name this field `signing_request` instead.

sergei-maertens commented on simple_certmanager/models.py:
this setup now messes with our expectations, every `Certificate` record so far was "guaranteed" to have a `public_certificate` file, but this is no longer the case now.

IMO the `Certificate` record should only be created when all the necessary information is available, and not any time sooner. To achieve this, I would set things up this way:

1. Keep the model `SigningRequest`, add the fields:
    * private_key (`models.TextField`), and make it `editable=False` - this is to hold the private key until we have the certificate and we need it to create the CSR
    * csr (`models.TextField`)
    * on save of this model -> generate the CSR if it doesn't exist yet, using the private key and entered information
2. Ensure that the CSR file can be downloaded from the admin interface, small convenience thing. You can use a Django `FileResponse` for this or something like that.
3. Add a form field to upload the certificate for an existing signing request (this does not need to be a model field, just a form field is enough)
4. Once the certificate is uploaded, validate that it matches the private key before proceeding (mistakes happen)
5. Once everything is valid, create the `Certificate` instance, copy the private key and certificate
6. add a FK/O2O field to the `SigningRequest` - `certificate` so that see which certificate was generated from the signing request. The CSR info may now be deleted if the user wishes to do so

CharString replied:
Oh nvm my remark about storing CSR then :)

sergei-maertens commented on simple_certmanager/templates/admin/simple_certmanager/change_list.html:
```diff
@@ -0,0 +1,8 @@
+{% load admin_list i18n %}
+
+{% block object-tools-items %}
+<li><a href="{% url 'generate_certificate' %}">Generate Certificate</a></li>
```
Can we not just use the standard admin views to add a certificate? Why do we need a custom view?

sergei-maertens commented on simple_certmanager/templates/simple_certmanager/certificatesigningrequestinfo_form.html:
same question as in `simple_certmanager/templates/admin/simple_certmanager/change_list.html`

sergei-maertens commented on simple_certmanager/validators.py:
```diff
@@ -46,3 +46,13 @@ class PrivateKeyValidator(PKIValidatorBase):
         load_pem_x509_private_key(file_content)
+
+
+class CertificateSigningRequestValidator(PKIValidatorBase):
+    message = _(
+        "Invalid file provided, expected a certificate signing request in PEM format"
+    )
+
+    @staticmethod
+    def validate(file_content: bytes) -> None:
+        load_pem_x509_csr(file_content)
```
This should not be needed since we generate the CSR, no?

sergei-maertens commented on testapp/settings.py:
```diff
@@ -62,3 +64,7 @@
 PRIVATE_MEDIA_ROOT = os.path.join(BASE_DIR, "private-media")
 PRIVATE_MEDIA_URL = "/private-media/"
+
+# FIXME : Replace this with a production-ready backend
```
you can drop this, it's just a test app

sergei-maertens commented on tests/test_admin.py:
```diff
@@ -64,6 +64,8 @@ def test_detail_view(temp_private_root, admin_client):
     html = response.content.decode("utf-8")
     doc = pq(html)
     uploads = doc(".file-upload")
+    print("Uploads:")
+    print(uploads.children())
```
please remove all print statements

sergei-maertens commented on tests/test_admin_views.py:
```diff
@@ -0,0 +1,129 @@
+from simple_certmanager.models import Certificate
+
+
+class GenerateCertificateViewTestCase(unittest.TestCase):
```
library tests should be in `pytest` style - we have a bunch of fixtures already in `tests/conftest.py` to generate private keys/certificates etc.

sergei-maertens commented on simple_certmanager/admin_views.py:
```diff
@@ -0,0 +1,145 @@
+    template_name_suffix = "_form"
+    model = CertificateSigningRequestInfo
+    form_class = CertificateSigningRequestInfoForm
+    success_url = "/admin/simple_certmanager/certificate/"
```
this will break as soon as someone puts the admin under a different URL in their project

sergei-maertens commented on simple_certmanager/admin_views.py:
```diff
@@ -0,0 +1,145 @@
+    def get_context_data(self, **kwargs):
+        return super().get_context_data(**kwargs)
+
+    def create_csr(
```
IMO this belongs in `simple_certmanager/utils.py` so that it can easily be unit tested and invoked from the `models.py`.

I think I'd even go further and create a dedicated module for it: `simple_certmanager/csr.py`

sergei-maertens commented on simple_certmanager/models.py:
```diff
@@ -16,6 +22,107 @@
+    common_name = models.CharField(
+        max_length=100, help_text=_("The common name for the certificate")
+    )
+    private_key = PrivateMediaFileField(
+        _("private key"),
+        help_text=_("The content of the private key"),
+        blank=True,
+        upload_to="ssl_certs_keys/%Y/%m/%d",
+        validators=[PrivateKeyValidator()],
+    )
```
I think we can avoid a file field entirey and just store the PEM data in a `models.TextField`, we don't need to have the file on disk to be able to generate a CSR.

The private key will be generated by us, this is not something that end users should upload.

sergei-maertens commented on simple_certmanager/models.py:
```diff
@@ -16,6 +22,107 @@
+    state_or_province_name = models.CharField(
+        max_length=100, help_text=_("The state or province name")
+    )
+    email_address = models.EmailField(help_text=_("Email address for the certificate"))
```
I don't think this should be a required field

sergei-maertens commented on simple_certmanager/models.py:
```diff
@@ -16,6 +22,107 @@
+    )
+    state_or_province_name = models.CharField(
+        max_length=100, help_text=_("The state or province name")
+    )
```
I don't think this should be a required field

sergei-maertens commented on simple_certmanager/models.py:
```diff
@@ -16,6 +22,107 @@
+    )
+    organization_name = models.CharField(
+        max_length=100, help_text=_("The name of the organization")
+    )
```
I don't think this should be a required field

sergei-maertens commented on simple_certmanager/models.py:
```diff
@@ -16,6 +22,107 @@
+class SigningRequest(DeleteFileFieldFilesMixin, models.Model):
+    country_name = models.CharField(
+        max_length=2, help_text=_("Two-letter country code")
+    )
```
I don't think this should be a required field

sergei-maertens commented on simple_certmanager/models.py:
```diff
@@ -16,6 +22,107 @@
+    def save(self, *args, **kwargs):
+        if not self.private_key:
+            # Generate private key
+            private_key = rsa.generate_private_key(
+                public_exponent=65537, key_size=2048, backend=default_backend()
+            )
```
@CharString - can you validate my thoughts?

If we're the ones generating the private keys, then:

1. Is RSA a sane default? I can't tell when you'd pick RSA/EC/DSA/EdDSA and why.
2. Key size 2048 for RSA seems on the lower end, I thought that 4096 is the go-to these days. Either way, I'd expect this to be configurable parameter with a default of 4096? Or at least a constant defined in `constants.py`
3. is it safe to use this hardcoded public exponent? I understand the risks are in the encryption itself when they don't properly apply padding, but we have to rely on openssl for this anyway

CharString replied:
1. RSA is the sanest IMHO, for our purposes. It's old, well studied and very simple (you can perform it with pen and paper with small enough keys). So the chance of bugs in implementation is smaller. The only reason I see to use elliptic curves is if space/compute is limited like on small devices, such as micro controllers.
2. +1 for putting key size in settings, so it's configurable depending on the use. 2048 bits is the bare minimum. 3072 bits strikes a nice middle ground between that and doubling it.
3. Yes, it's safe. It's part of the public key and needs to be relative prime. Too large is not only computationally expensive, but actively bad. And yes, proper padding solves issues with the smaller primes.

sergei-maertens commented on simple_certmanager/models.py:
```diff
@@ -16,6 +22,107 @@
+            )
+            private_key_file_bytes = private_key.private_bytes(
+                encoding=serialization.Encoding.PEM,
+                format=serialization.PrivateFormat.PKCS8,
```
why this format and not another? The utilities/tests have been using traditional openssl

sergei-maertens commented on simple_certmanager/models.py:
```diff
@@ -16,6 +22,107 @@
+                format=serialization.PrivateFormat.PKCS8,
+                encryption_algorithm=serialization.NoEncryption(),
+            )
+            private_key_file = ContentFile(
+                private_key_file_bytes, name="private_key.pem"
+            )
+            self.private_key = private_key_file
```
when using a plain `models.TextField`, you can store it as `private_key_file_bytes.decode("ascii")`. It's all base64 so should be fine to store as ascii

CharString replied:
PEM files are base64, yes. But keep in mind that the key itself are arbitrary bytes and may not decode as ASCII to text.

sergei-maertens replied:
yeah we only deal with PEM - the way the certificates are used requires PEM and while we can transform between them, I don't think we should invite that extra layer of complexity

sergei-maertens commented on simple_certmanager/models.py:
```diff
@@ -16,6 +22,107 @@
+    _private_key_obj: PrivateKeyTypes | None = None
+
+    def save(self, *args, **kwargs):
+        if not self.private_key:
```
I think all this code here belongs in a utility function or class method rather than implemented directly in the `save`, that would make it easier to unit-test.

sergei-maertens commented on simple_certmanager/models.py:
```diff
@@ -16,6 +22,107 @@
+            )
+
+            if self.email_address:
+                csr_builder = csr_builder.add_extension(
+                    x509.SubjectAlternativeName([x509.RFC822Name(self.email_address)]),
+                    critical=False,
+                )
```
does the email address really go in a SAN extension field? I've only ever seen DNS names and IP addresses in there

CharString replied:
Depends on what the subject is.. if the certificate identifies a person it can be an email address.
It can be used as an alternative to PGP for email for instance... But in our usecase it's probably a service or organisation. For the former, URI or hostnames make more sense.

sergei-maertens commented on simple_certmanager/models.py:
```diff
@@ -16,6 +22,107 @@
+
+        super().save(*args, **kwargs)
+
+    @property
+    def _private_key(self):
+        if self._private_key_obj is None:
+            with self.private_key.open(mode="rb") as key_f:
+                self._private_key_obj = load_pem_x509_private_key(key_f.read())
+        return self._private_key_obj
```
this seems unused

sergei-maertens commented on simple_certmanager/models.py:
```diff
@@ -43,6 +150,13 @@ class Certificate(DeleteFileFieldFilesMixin, models.Model):
         upload_to="ssl_certs_keys/%Y/%m/%d",
         validators=[PrivateKeyValidator()],
     )
+    signing_request = models.OneToOneField(
+        SigningRequest,
+        on_delete=models.SET_NULL,
+        null=True,
+        blank=True,
+        related_name="certificate",
+    )
```
I would make this relation on the opposite side, I'm more interested in which certificate was created from a CSR than from which CSR a certificate comes.

That also makes it that I can delete the CSRs once I have received the certificate and clean up any intermediate data, without breaking anything.

sergei-maertens commented on testapp/urls.py:
```diff
@@ -2,15 +2,19 @@
+            ("simple_certmanager.admin_urls", "simple_certmanager"),
+            namespace="simple_certmanager",
+        ),
+    ),
```
I don't think you need to define any custom URLs to be included, instead you can just override `get_urls` on the model admin class

sergei-maertens commented on testapp/urls.py:
```diff
@@ -2,15 +2,19 @@
 
-# NOTE: The staticfiles_urlpatterns also discovers static files (ie. no need
-# to run collectstatic). Both the static folder and the media folder are only
-# served via Django if DEBUG = True.
```
bit rude to delete comments that explain things :)

sergei-maertens commented on tests/conftest.py:
```diff
@@ -32,6 +32,22 @@ def root_cert(root_key) -> x509.Certificate:
 
 
+@pytest.fixture(scope="session")
+def root_csr(root_key) -> x509.CertificateSigningRequest:
```
I don't see this fixture used anywhere?

sergei-maertens commented on tests/data/csr.pem:
We should not add new PEM files for certificates for tests. They expire, and it's better to use the pytest fixtures to generate the stuff we need on-the-fly.

sergei-maertens commented on tests/test_admin.py:
```diff
@@ -37,6 +45,30 @@ def test_list_view(temp_private_root, admin_client):
+    fields = doc(".field-common_name")
+    anchor = fields[0].getchildren()[0]
+    assert anchor.tag == "a"
+    assert anchor.text == csr.common_name
```
common name can be anything, e.g. "Gemeente Amsterdam". I don't think it makes sense to display this in an anchor tag

sergei-maertens commented on tests/test_admin.py:
```diff
@@ -37,6 +45,30 @@ def test_list_view(temp_private_root, admin_client):
     assert anchor.text == certificate.label
 
 
+def test_csr_list_view(temp_private_root, admin_client):
```
can you create a new module for the CSR model? `test_signing_request_admin.py` and put the new tests there.

sergei-maertens commented on tests/test_admin.py:
```diff
@@ -78,6 +110,149 @@ def test_detail_view(temp_private_root, admin_client):
+    )
+
+
+def test_adding_csr_generates_CSR_and_PK(temp_private_root, admin_client):
```
a better test is using `admin_client.post(...)` to submit the form field data and then check the created record for the expected state - you want to test what users will be doing, which is:

1. navigate to the admin
2. click the add button
3. fill out the forms
4. save it
5. download the CSR file

sergei-maertens commented on tests/test_admin.py:
```diff
@@ -78,6 +110,149 @@ def test_detail_view(temp_private_root, admin_client):
+    form = SigningRequestAdminForm(
+        instance=csr, data=form_data, files={"certificate": invalid_cert_pem}
+    )
+    assert not form.is_valid()
```
I'm more interested in the actual file upload and validation errors displayed, using `admin_client.post(...)`

The form is an implementation detail, it gets tested through the integration test(s)

sergei-maertens commented on tests/test_admin.py:
```diff
@@ -78,6 +110,149 @@ def test_detail_view(temp_private_root, admin_client):
+    assert not form.is_valid()
+
+    model_admin = SigningRequestAdmin(SigningRequest, admin.site)
+    model_admin.save_model(admin_client, csr, form, change=False)
```
testing implementation details way too much, please use only the `admin_client` :)

sergei-maertens commented on simple_certmanager/forms.py:
```diff
@@ -17,3 +21,54 @@ def __init__(self, *args, **kwargs):
+        fields = "__all__"
+
+    def __init__(self, *args, **kwargs):
+        super(SigningRequestAdminForm, self).__init__(*args, **kwargs)
```
```suggestion
        super().__init__(*args, **kwargs)
```

CharString replied:
Co-pilot? Is that you?

sergei-maertens commented on simple_certmanager/forms.py:
```diff
@@ -17,3 +21,54 @@ def __init__(self, *args, **kwargs):
+        # Check if private key and CSR are empty
+        private_key_empty = not self.instance.private_key
+        csr_empty = not self.instance.csr
+        if private_key_empty or csr_empty:
+            self.fields["certificate"].widget = forms.HiddenInput()
```
better even:

```py
del self.fields["certificate"]
```

sergei-maertens commented on simple_certmanager/forms.py:
```diff
@@ -17,3 +21,54 @@ def __init__(self, *args, **kwargs):
+        if private_key_empty or csr_empty:
+            self.fields["certificate"].widget = forms.HiddenInput()
+        else:
+            self.fields["certificate"].widget = forms.FileInput()
```
not needed, the file input is the default widget for a file field, no?

sergei-maertens commented on simple_certmanager/forms.py:
```diff
@@ -17,3 +21,54 @@ def __init__(self, *args, **kwargs):
+            self.fields["certificate"].widget = forms.FileInput()
+
+    def clean(self):
+        cleaned_data = super().clean()
```
you should use `self.cleaned_data` instead, there's no guarantee about the return value of `clean`

sergei-maertens commented on simple_certmanager/forms.py:
```diff
@@ -17,3 +21,54 @@ def __init__(self, *args, **kwargs):
+        certificate_file = cleaned_data.get("certificate")
+        if certificate_file:
+            try:
+                certificate_data = certificate_file.read()
+                certificate = x509.load_pem_x509_certificate(certificate_data)
+                csr = x509.load_pem_x509_csr(self.instance.csr.encode())
+                # Check if the certificate matches the CSR
+                public_key = csr.public_key()
+                public_key.verify(
+                    certificate.signature,
+                    certificate.tbs_certificate_bytes,
+                    padding.PKCS1v15(),
+                    certificate.signature_hash_algorithm,
+                )
```
@CharString does this look right?

This padding argument seems... arbitrary?

CharString replied:
This may not be correct. The same padding scheme should be used during verification as the issuer used during signing. That can vary from issuer to issuer. If asking the user is unfeasible (the library is called "simple" after all :)) we might be able to do heuristics on the cipher. Depending on if PKCS 1.5 or the newer version is used, the cipher may look recognisably different.

sergei-maertens commented on simple_certmanager/forms.py:
```diff
@@ -17,3 +21,54 @@ def __init__(self, *args, **kwargs):
+                    "certificate",
+                    "Certificate does not match the signature from the actual CSR.",
+                )
+            except Exception as e:
+                self.add_error("certificate", f"An unexpected error occurred: {str(e)}")
```
do not ever dump tracebacks to end-users! this can leak sensitive information.

sergei-maertens commented on simple_certmanager/admin_urls.py:
this... doesn't do anything? :stuck_out_tongue:

sergei-maertens commented on simple_certmanager/admin.py:
```diff
@@ -23,7 +112,10 @@ class CertificateAdmin(PrivateMediaMixin, admin.ModelAdmin):
     list_filter = ("type",)
     search_fields = ("label", "type")
-    readonly_fields = ("serial_number",)
+    readonly_fields = (
+        "serial_number",
+        "signing_request",
+    )
```
I don't think it's useful to have this field here

sergei-maertens commented on simple_certmanager/admin.py:
```diff
@@ -1,19 +1,108 @@
+    search_fields = ("common_name", "organization_name")
+    private_media_fields = ("private_key",)
+    private_media_no_download_fields = ("private_key",)
+    actions = [download_csr]
```
neat

I'd also expect a link on the detail page and in the admin changelist page for each record

sergei-maertens commented on simple_certmanager/admin.py:
```diff
@@ -1,19 +1,108 @@
+
+    def save_model(self, request, obj, form, change):
+        super().save_model(request, obj, form, change)
+        validated_certificate = form.cleaned_data.get("certificate")
+        if validated_certificate:
+            created_certificate = obj.create_certificate(validated_certificate)
+            return HttpResponseRedirect(
+                reverse(
+                    "admin:simple_certmanager_certificate_change",
+                    args=[created_certificate.pk],
+                )
+            )
```
what happens when (another) certificate is uploaded to the same CSR?

sam-bertin committed:
CSR Generation + Storage

Updated Documentation

Info isn't required from Admin

Missing Dependecy

Public Certificate Not Required  for AdminFrom + Tests

Public certificate IS required. Update Tests

More tests for better coverage

Versioning

Review pt.1

sam-bertin commented:
Let's close this one.
It has been cut in #39 and #40
